tests/MongoDB.Driver.Core.Tests/Core/Operations/AsyncCursorTests.cs

```diff
@@ -313,143 +311,151 @@ private ConnectionDescription CreateConnectionDescriptionSupportingSession()
+            Insert(documents);
 
-            // constructors
-            public Reflector(AsyncCursor<BsonDocument> instance)
+            var findOperation = new FindOperation<BsonDocument>(_collectionNamespace, BsonDocumentSerializer.Instance, _messageEncoderSettings)
             {
-                _instance = instance;
-            }
+                BatchSize = batchSize
+            };
 
-            // public properties
-            public int? BatchSize
+            using (var binding = new WritableServerBinding(CoreTestConfiguration.Cluster, _session.Fork()))
+            using (var cursor = findOperation.Execute(binding, CancellationToken.None))
             {
-                get
+                AssertExpectedChannelSourceDisposedValue(cursor);
+                while (cursor.MoveNext())
                 {
-                    var fieldInfo = _instance.GetType().GetField("_batchSize", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (int?)fieldInfo.GetValue(_instance);
+                    AssertExpectedChannelSourceDisposedValue(cursor);
                 }
+                AssertExpectedChannelSourceDisposedValue(cursor);
             }
+        }
 
-            public IChannelSource ChannelSource
-            {
-                get
-                {
-                    var fieldInfo = _instance.GetType().GetField("_channelSource", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (IChannelSource)fieldInfo.GetValue(_instance);
-                }
-            }
+        // private methods
+        private void AssertExpectedChannelSourceDisposedValue(IAsyncCursor<BsonDocument> cursor)
+        {
+            var cursorImplementation = (AsyncCursor<BsonDocument>)cursor;
+            var channelSource = cursorImplementation._channelSource();
+            var cursorId = cursorImplementation._cursorId();
+            var actualDisposed = GetDisposed(channelSource);
+            var expectedDisposed = cursorId == 0;
+            actualDisposed.Should().Be(expectedDisposed);
+        }
 
-            public CollectionNamespace CollectionNamespace
+        private bool GetDisposed(IChannelSource channelSource)
+        {
+            if (channelSource == null)
             {
-                get
-                {
-                    var fieldInfo = _instance.GetType().GetField("_collectionNamespace", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (CollectionNamespace)fieldInfo.GetValue(_instance);
-                }
+                return true;
             }
-
-            public int Count
+            else
             {
-                get
-                {
-                    var fieldInfo = _instance.GetType().GetField("_count", BindingFlags.NonPublic | BindingFlags.Instance);
-                    return (int)fieldInfo.GetValue(_instance);
-                }
+                var type = channelSource.GetType();
```

Comment:
This is annoying. Normally when we do reflection we KNOW the type we are reflecting over. But because of all the layers of wrapping we're not quite sure what type the channel source is, so we have to use the actual type and ASSUME that whatever type it turned out to be has a field named "_disposed".

Reply:
Yeah, it's a bit ugly.  I gather there's no Disposed property to just tell you whether it is or not.